tket/src/NumericalOptimiser/NumericalOptimiser.cpp

```diff
@@ -0,0 +1,108 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "NumericalOptimiser.hpp"
+
+namespace tket {
+
+Circuit optimise(Circuit &circ, Architecture &arch, unsigned int k) {
+  PartitionVec post_synthesis;
+  PartitionVec pre_synthesis = partition(circ, arch, k);
+
+  for (Partition partition : pre_synthesis) {
+    post_synthesis.insert(post_synthesis.begin(), synthesise(partition, arch));
+  }
+
+  for (Partition partition : post_synthesis) {
+    EdgeVec edges = {};
+    for (Qubit qubit : partition.second) {
```

Comment:
I'm a bit confused about this. It seems that the `to_replace` subcircuit is empty (the input edges are the same as the output edges, no gates in between); so we are just prepending the replacements to the original circuit, without removing anything. Am I misunderstanding?

Reply:
Partition() has the side effect of removing each partition so the circuit is empty when it comes time to add the re-synthesised sub circuits